### PR TITLE
Allow services to join an existing transaction

### LINKED ISSUE #490: Improve transaction control in service layer
Currently, it is problematic for transactional service methods to invoke other transactional services, as transaction control is managed within each service operation manually on an "as-needed" basis, using the `Transactional()` function.

There should be a simpler way for services to define transaction boundaries, and to allow the invocation of other service methods within the scope of the same transaction.

At present, transactional code looks like this:

```
err := transaction.Transactional(s.tm, func(tr transaction.TransactionalResources) error {
  // do something
  return err
}
```

We should be able to simplify this further, perhaps by providing a `Transactional()` method and transaction management features in a base service class.

### PR DESCRIPTION
**Pull request by sbryzak**
Fixes #490

### CODE REVIEW

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-496-1`

File: application/service/factory/service_factory.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+				return nil
+			case <-txTimeout:
+				log.Debug(nil, nil, "Rolling back the transaction...")
+				tx.Rollback()
+				log.Error(nil, nil, "database transaction timeout!")
+				return errors.New("database transaction timeout!")
+			}
+		}()
+
+		return err
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply application/service/factory/service_factory.go]: Done in https://github.com/fabric8-services/fabric8-auth/pull/496/commits/8f1ae4bc24d03a56aa0a390bc4eefed4e48a9231

File: application/service/factory/service_factory.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			return errors.WithStack(err)
+		}
+
+		// Set the transaction flag to true
+		s.inTransaction = true
+
+		// Set the transactional repositories property
+		s.transactionalRepositories = tx.(repository.Repositories)
+
+		// Ensure changes are reverted at the end of the transaction
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply application/service/factory/service_factory.go]: Done in https://github.com/fabric8-services/fabric8-auth/pull/496/commits/8f1ae4bc24d03a56aa0a390bc4eefed4e48a9231

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-496-2`

**Pull request merged**